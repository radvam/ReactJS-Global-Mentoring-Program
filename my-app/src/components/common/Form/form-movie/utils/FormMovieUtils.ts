import * as yup from "yup";

export const genresData = [
  { value: "Crime", id: 1 },
  { value: "Documentary", id: 2 },
  { value: "Horror", id: 3 },
  { value: "Comedy", id: 4 },
  { value: "Fantasy", id: 5 },
  { value: "Animation", id: 6 },
  { value: "Family", id: 7 },
  { value: "Science Fiction", id: 8 },
  { value: "Action", id: 9 },
  { value: "Drama", id: 10 },
  { value: "Romance", id: 11 },
  { value: "Thriller", id: 12 },
  { value: "Adventure", id: 13 },
  { value: "History", id: 14 },
  { value: "Mystery", id: 15 },
  { value: "Music", id: 16 },
  { value: "War", id: 17 },
];

export const formAddInitialValues = {
  title: "",
  release_date: "",
  poster_path: "",
  genres: [],
  overview: "",
  runtime: 0,
};

export const validationSchema = yup.object().shape({
  title: yup.string().required(),
  release_date: yup.date().required(),
  poster_path: yup.string().url().required(),
  genres: yup.array().defined().min(1).required(),
  overview: yup.string().required(),
  runtime: yup.number().min(1).required().integer(),
  id: yup.number(),
});
