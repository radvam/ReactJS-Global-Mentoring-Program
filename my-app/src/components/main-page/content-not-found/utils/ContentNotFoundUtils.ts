export const refreshPage = (): void => {
  window.location.reload();
};
