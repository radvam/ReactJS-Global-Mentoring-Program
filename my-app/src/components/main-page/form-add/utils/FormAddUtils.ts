export const selectGenresData = [
  { value: "gold", label: "Crime" },
  { value: "lime", label: "Documentary" },
  { value: "green", label: "Horror" },
  { value: "cyan", label: "Comedy" },
  { value: "red", label: "Fantasy" },
  { value: "blue", label: "Animation" },
  { value: "lightSalmon", label: "Family" },
  { value: "brown", label: "Science Fiction" },
  { value: "crimson", label: "Action" },
  { value: "lightCoral", label: "Drama" },
  { value: "teal", label: "Romance" },
  { value: "silver", label: "Thriller" },
  { value: "slateGray", label: "Adventure" },
  { value: "darkSlateGrey", label: "History" },
  { value: "yellow", label: "Mystery" },
  { value: "rosyBrown", label: "Music" },
  { value: "bisque", label: "War" },
];

export const defineColorsByGenres = (genres: string[]): string[] => {
  const result: string[] = [];
  genres.forEach((item) => {
    selectGenresData.forEach(({ label, value }) => {
      if (label === item) {
        result.push(value);
      }
    });
  });
  return result;
};

export const defineGenresByColors = (array: string[]): string[] => {
  const result: string[] = [];
  array.forEach((item) => {
    selectGenresData.forEach(({ label, value }) => {
      if (value === item) {
        result.push(label);
      }
    });
  });
  return result;
};
