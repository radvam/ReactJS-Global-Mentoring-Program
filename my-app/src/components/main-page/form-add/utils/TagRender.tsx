import React from "react";
import { Tag } from "antd";


export const TagRender = (props: Record<string, any>): React.ReactElement => {
  const { label, value, closable, onClose } = props;

  return (
    <Tag
      color={value}
      closable={closable}
      onClose={onClose}
      style={{ marginRight: 3 }}
    >
      {label}
    </Tag>
  );
};
