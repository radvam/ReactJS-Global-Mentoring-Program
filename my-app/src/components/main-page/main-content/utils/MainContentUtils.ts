import { Card } from "../MainContent.interface";

const cardListData: Card[] = [
  {
    id: 338970,
    name: "Tomb Raider",
    date: "2018",
    imageSrc: "https://image.tmdb.org/t/p/w500/ePyN2nX9t8SOl70eRW47Q29zUFO.jpg",
    genres: "Action, Adventure",
  },
  {
    id: 337167,
    name: "Fifty Shades Freed",
    date: "2018",
    imageSrc: "https://image.tmdb.org/t/p/w500/3kcEGnYBHDeqmdYf8ZRbKdfmlUy.jpg",
    genres: "Drama, Romance",
  },
  {
    id: 353616,
    name: "Pitch Perfect 3",
    date: "2017",
    imageSrc: "https://image.tmdb.org/t/p/w500/fchHLsLjFvzAFSQykiMwdF1051.jpg",
    genres: "Comedy, Music",
  },
  {
    id: 284053,
    name: "Thor: Ragnarok",
    date: "2017",
    imageSrc: "https://image.tmdb.org/t/p/w500/rzRwTcFvttcN1ZpX2xv4j3tSdJu.jpg",
    genres: "Action, Adventure, Fantasy",
  },
  {
    id: 345923,
    name: "All I See Is You",
    date: "2017",
    imageSrc: "https://image.tmdb.org/t/p/w500/wc9MgFWuBl3VS72w6VtiZXTgN79.jpg",
    genres: "Drama, Thriller",
  },
  {
    id: 321612,
    name: "Beauty and the Beast",
    date: "2017",
    imageSrc: "https://image.tmdb.org/t/p/w500/tWqifoYuwLETmmasnGHO7xBjEtt.jpg",
    genres: "Family, Fantasy, Romance",
  },
  {
    id: 399055,
    name: "The Shape of Water",
    date: "2017",
    imageSrc: "https://image.tmdb.org/t/p/w500/k4FwHlMhuRR5BISY2Gm2QZHlH5Q.jpg",
    genres: "Drama, Fantasy, Romance",
  },
  {
    id: 268896,
    name: "Pacific Rim: Uprising",
    date: "2018",
    imageSrc: "https://image.tmdb.org/t/p/w500/v5HlmJK9bdeHxN2QhaFP1ivjX3U.jpg",
    genres: "Action, Fantasy, Science Fiction, Adventure",
  },
  {
    id: 335984,
    name: "Blade Runner 2049",
    date: "2017",
    imageSrc: "https://image.tmdb.org/t/p/w500/gajva2L0rPYkEWjzgFlBXCAVBE5.jpg",
    genres: "Mystery, Science Fiction, Thriller",
  },
  {
    id: 353486,
    name: "Jumanji: Welcome to the Jungle",
    date: "2017",
    imageSrc: "https://image.tmdb.org/t/p/w500/bXrZ5iHBEjH7WMidbUDQ0U2xbmr.jpg",
    genres: "Action, Adventure, Comedy, Family",
  },
  {
    id: 10345,
    name: "K-9",
    date: "1989",
    imageSrc: "https://image.tmdb.org/t/p/w500/4SmGgyoIRzOqLI9M5lof2e00B16.jpg",
    genres: "Action, Comedy, Crime",
  },
];

export default cardListData;
