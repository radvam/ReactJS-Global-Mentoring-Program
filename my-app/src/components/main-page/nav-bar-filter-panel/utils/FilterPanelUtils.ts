import { FilterDataItem } from "../FilterPanel.interface";

export const filterData: FilterDataItem[] = [
  { id: "1", name: "ALL" },
  { id: "2", name: "DOCUMENTARY" },
  { id: "3", name: "COMEDY" },
  { id: "4", name: "HORROR" },
  { id: "5", name: "CRIME" },
];
