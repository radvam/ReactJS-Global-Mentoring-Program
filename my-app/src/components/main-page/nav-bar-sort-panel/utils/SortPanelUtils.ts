import { SelectDataItem } from "../SortPanel.interface";

const selectData: SelectDataItem[] = [
  { id: "1", value: "RELEASE DATE" },
  { id: "2", value: "SECOND" },
  { id: "3", value: "THIRD" },
];

export default selectData;
