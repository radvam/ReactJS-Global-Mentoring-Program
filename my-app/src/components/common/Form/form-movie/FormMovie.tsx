import React, { FC } from "react";
import { Formik } from "formik";

import { genresData, validationSchema } from "./utils/FormMovieUtils";

import { FormMovieProps } from "./FormMovie.interface";

import {
  ID_LABEL,
  TITLE_LABEL,
  DATE_LABEL,
  URL_LABEL,
  GENRE_LABEL,
  OVERVIEW_LABEL,
  RUNTIME_LABEL,
} from "../../../../constants/mainPageConstants";

import {
  FormMovieWrapper,
  FormItem,
  Title,
  Error,
  ButtonGroup,
  Submit,
  Reset,
  MovieId,
} from "./FormMovie.style";

export const FormMovie: FC<FormMovieProps> = ({
  setOpenModal,
  onSubmit,
  formInitialValues,
  resetButtonText,
  submitButtonText,
  id,
}): React.ReactElement => {
  return (
    <Formik
      initialValues={formInitialValues}
      validationSchema={validationSchema}
      onSubmit={async (values, { resetForm }) => {
        setOpenModal(false);
        await onSubmit(values);
        resetForm({});
      }}
    >
      {({
        values,
        errors,
        touched,
        handleChange,
        handleReset,
        handleBlur,
        handleSubmit,
        isSubmitting,
      }) => (
        <form onSubmit={handleSubmit}>
          <FormMovieWrapper>
            {id && (
              <FormItem>
                <Title>{ID_LABEL}</Title>
                <MovieId>{id}</MovieId>
              </FormItem>
            )}
            <FormItem>
              <Title>{TITLE_LABEL}</Title>
              {errors.title && touched.title && <Error>{errors.title}</Error>}
              <input
                type="text"
                name="title"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.title}
                placeholder={"Title here"}
              />
            </FormItem>
            <FormItem>
              <Title>{DATE_LABEL}</Title>
              {errors.release_date && touched.release_date && (
                <Error>{errors.release_date}</Error>
              )}
              <input
                type="date"
                name="release_date"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.release_date}
              />
            </FormItem>
            <FormItem>
              <Title>{URL_LABEL}</Title>
              {errors.poster_path && touched.poster_path && (
                <Error>{errors.poster_path}</Error>
              )}
              <input
                type="text"
                name="poster_path"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.poster_path}
                placeholder={"Movie URL here"}
              />
            </FormItem>
            <FormItem>
              <Title>{GENRE_LABEL}</Title>
              {errors.genres && touched.genres && (
                <Error>{errors.genres}</Error>
              )}
              <select
                name="genres"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.genres}
                multiple
              >
                {genresData.map(({ value }) => (
                  <option value={value}>{value}</option>
                ))}
              </select>
            </FormItem>
            <FormItem>
              <Title>{OVERVIEW_LABEL}</Title>
              {errors.overview && touched.overview && (
                <Error>{errors.overview}</Error>
              )}
              <input
                type="text"
                name="overview"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.overview}
                placeholder={"Overview here"}
              />
            </FormItem>
            <FormItem>
              <Title>{RUNTIME_LABEL}</Title>
              {errors.runtime && touched.runtime && (
                <Error>{errors.runtime}</Error>
              )}
              <input
                type="number"
                name="runtime"
                onChange={handleChange}
                onBlur={handleBlur}
                value={values.runtime}
                min={0}
              />
            </FormItem>
            <ButtonGroup>
              <Reset type="reset" onClick={handleReset}>
                {resetButtonText}
              </Reset>
              <Submit type="submit" disabled={isSubmitting}>
                {submitButtonText}
              </Submit>
            </ButtonGroup>
          </FormMovieWrapper>
        </form>
      )}
    </Formik>
  );
};
