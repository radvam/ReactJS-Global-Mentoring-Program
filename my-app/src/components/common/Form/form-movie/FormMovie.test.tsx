import * as React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";

import { FormMovie } from "./FormMovie";
import { formAddInitialValues } from "./utils/FormMovieUtils";

import { ID_LABEL } from "../../../../constants/mainPageConstants";

const newMovie = {
  title: "Tomb Raider 5",
  release_date: "2021-26-04",
  poster_path:
    "https://image.tmdb.org/t/p/w500/ePyN2nX9t8SOl70eRW47Q29zUFO.jpg",
  overview:
    "Lara Croft, the fiercely independent daughter of a missing adventurer, must push herself beyond her limits when she finds herself on the island where her father disappeared.",
  genres: ["Action", "Adventure"],
  runtime: 118,
};

const onSubmit = jest.fn();
const setOpenModal = jest.fn();
const submitButtonText = "Submit";
const resetButtonText = "Reset";

describe("FormMovie:", () => {
  test("should initialize", () => {
    const { asFragment } = render(
      <FormMovie
        setOpenModal={setOpenModal}
        onSubmit={onSubmit}
        formInitialValues={formAddInitialValues}
        submitButtonText={submitButtonText}
        resetButtonText={resetButtonText}
      />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("should be reset form when click Reset button", () => {
    render(
      <FormMovie
        setOpenModal={setOpenModal}
        onSubmit={onSubmit}
        formInitialValues={formAddInitialValues}
        submitButtonText={submitButtonText}
        resetButtonText={resetButtonText}
      />
    );

    const nameTextField = screen.getByTestId("title");
    const resetButton = screen.getByTestId("reset");
    const changeTitle = "change title";

    fireEvent.change(nameTextField, {
      target: {
        value: changeTitle,
      },
    });

    expect(nameTextField.getAttribute("value")).toEqual(changeTitle);

    fireEvent.click(resetButton);

    expect(nameTextField.getAttribute("value")).toEqual(
      formAddInitialValues.title
    );
  });

  test("should be validate form when slick Submit button", async () => {
    render(
      <FormMovie
        setOpenModal={setOpenModal}
        onSubmit={onSubmit}
        formInitialValues={formAddInitialValues}
        submitButtonText={submitButtonText}
        resetButtonText={resetButtonText}
      />
    );
    const submitButton = screen.getByTestId("submit");

    fireEvent.click(submitButton);

    await waitFor(async () => {
      const message = await screen.findByText("title is a required field");
      expect(message).toBeTruthy();
    });
  });

  test("should be render id form item", async () => {
    render(
      <FormMovie
        setOpenModal={setOpenModal}
        onSubmit={onSubmit}
        formInitialValues={formAddInitialValues}
        submitButtonText={submitButtonText}
        resetButtonText={resetButtonText}
        id={12345}
      />
    );

    const expectedItem = screen.getByText(ID_LABEL);

    expect(expectedItem).toBeInTheDocument();
  });

  test("should be called functions by Submit", async () => {
    render(
      <FormMovie
        setOpenModal={setOpenModal}
        onSubmit={onSubmit}
        formInitialValues={newMovie}
        submitButtonText={submitButtonText}
        resetButtonText={resetButtonText}
      />
    );

    const handleSubmit = () => {
      setOpenModal();
      onSubmit();
    };

    const form = screen.getByTestId("form");
    form.onsubmit = handleSubmit;

    fireEvent.submit(form);

    expect(setOpenModal).toHaveBeenCalled();
    expect(onSubmit).toHaveBeenCalled();
  });
});
