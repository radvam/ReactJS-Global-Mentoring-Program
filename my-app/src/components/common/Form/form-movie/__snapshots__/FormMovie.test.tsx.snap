// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormMovie: should initialize 1`] = `
<DocumentFragment>
  <form
    data-testid="form"
  >
    <div
      class="sc-bdfBwQ dtwnJa"
    >
      <div
        class="sc-gsTCUz hFmoFt"
      >
        <p
          class="sc-dlfnbm icrUWe"
        >
          Title
        </p>
        <input
          data-testid="title"
          name="title"
          placeholder="Title here"
          type="text"
          value=""
        />
      </div>
      <div
        class="sc-gsTCUz hFmoFt"
      >
        <p
          class="sc-dlfnbm icrUWe"
        >
          Release Date
        </p>
        <input
          name="release_date"
          type="date"
          value=""
        />
      </div>
      <div
        class="sc-gsTCUz hFmoFt"
      >
        <p
          class="sc-dlfnbm icrUWe"
        >
          Movie Url
        </p>
        <input
          name="poster_path"
          placeholder="Movie URL here"
          type="text"
          value=""
        />
      </div>
      <div
        class="sc-gsTCUz hFmoFt"
      >
        <p
          class="sc-dlfnbm icrUWe"
        >
          Genre
        </p>
        <select
          multiple=""
          name="genres"
        >
          <option
            value="Crime"
          >
            Crime
          </option>
          <option
            value="Documentary"
          >
            Documentary
          </option>
          <option
            value="Horror"
          >
            Horror
          </option>
          <option
            value="Comedy"
          >
            Comedy
          </option>
          <option
            value="Fantasy"
          >
            Fantasy
          </option>
          <option
            value="Animation"
          >
            Animation
          </option>
          <option
            value="Family"
          >
            Family
          </option>
          <option
            value="Science Fiction"
          >
            Science Fiction
          </option>
          <option
            value="Action"
          >
            Action
          </option>
          <option
            value="Drama"
          >
            Drama
          </option>
          <option
            value="Romance"
          >
            Romance
          </option>
          <option
            value="Thriller"
          >
            Thriller
          </option>
          <option
            value="Adventure"
          >
            Adventure
          </option>
          <option
            value="History"
          >
            History
          </option>
          <option
            value="Mystery"
          >
            Mystery
          </option>
          <option
            value="Music"
          >
            Music
          </option>
          <option
            value="War"
          >
            War
          </option>
        </select>
      </div>
      <div
        class="sc-gsTCUz hFmoFt"
      >
        <p
          class="sc-dlfnbm icrUWe"
        >
          Overview
        </p>
        <input
          name="overview"
          placeholder="Overview here"
          type="text"
          value=""
        />
      </div>
      <div
        class="sc-gsTCUz hFmoFt"
      >
        <p
          class="sc-dlfnbm icrUWe"
        >
          Runtime
        </p>
        <input
          min="0"
          name="runtime"
          type="number"
          value="0"
        />
      </div>
      <div
        class="sc-eCssSg jIHwOa"
      >
        <button
          class="sc-jSgupP sc-gKsewC jWwiRE iwWjtp"
          data-testid="reset"
          name="reset"
          type="reset"
        >
          Reset
        </button>
        <button
          class="sc-jSgupP jWwiRE"
          data-testid="submit"
          type="submit"
        >
          Submit
        </button>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
